### PR TITLE
feat: set data structure

### PR DESCRIPTION
**Pull request by enitrat**
## Pull Request type

Implements an array-based `Set` data structure (used for AccessList)

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves: #

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/utils/src/set.cairo [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    #[inline]
+    fn len(self: @Set<T>) -> usize {
+        self.inner.span().len()
+    }
+}
+
+impl SetTCloneImpl<T, +Clone<T>, +Drop<T>, +PartialEq<T>, +Copy<T>> of Clone<Set<T>> {
+    fn clone(self: @Set<T>) -> Set<T> {
+        let mut response: Array<T> = Default::default();
```
Comment (ClementWalter) [inline crates/utils/src/set.cairo:69]: "response" ?

File: crates/utils/src/set.cairo [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        self.inner.contains(item)
+    }
+
+    #[inline]
+    fn to_span(self: SpanSet<T>) -> Span<T> {
+        self.inner
+    }
+
+    fn clone_set(self: SpanSet<T>) -> Set<T> {
+        let mut response: Array<T> = Default::default();
```
Comment (ClementWalter) [inline crates/utils/src/set.cairo:111]: "response" (surprised by the naming)

File: crates/utils/src/helpers.cairo [hunk 3]
```diff
@@ -514,6 +523,20 @@ impl SpanExtension<T, +Copy<T>, +Drop<T>> of SpanExtTrait<T> {
             }
         }
     }
+
+    // Returns the index of an item in the array.
+    fn index_of<+PartialEq<T>>(mut self: Span<T>, value: T) -> u128 {
```
Comment (ClementWalter) [inline crates/utils/src/helpers.cairo:528]: `Option<u128>` ?

File: crates/utils/src/set.cairo (hunk 1 above)
Comment (enitrat) [reply crates/utils/src/set.cairo:69]: nomenclature from the corelib

File: crates/utils/src/helpers.cairo (hunk 3 above)
Comment (enitrat) [reply crates/utils/src/helpers.cairo]: Fixed

**Review by ClementWalter — DISMISSED** [review]

**Pull request merged**